lib/proxy_ssh/proxy.go

```diff
@@ -0,0 +1,282 @@
+				if request.Type == "exit-status" {
+					// Connection is closed, end the proxy loop and remove this
+					// session from the list of known sessions.
+					p.sessions.Delete(remoteAddrString)
+					// End the proxy loop
+					break
+				}
+			}
+
+			// End the communication between the local and remote.
+			localChannel.Close()
+			remoteChannel.Close()
+		}()
+
+		// Connect the channels.
+		log.Debugf("Begin session connection between %q and %q.", conn.RemoteAddr(), remoteConn.RemoteAddr())
+		go io.Copy(remoteChannel, localChannel)
+		go io.Copy(localChannel, remoteChannel)
+		// These are kept for safety to ensure the channels are indeed closed,
+		// but in theory the ProxyLoop will close the channels and that will
+		// signal to io.Copy that we are complete.
+		defer localChannel.Close()
+		defer remoteChannel.Close()
+	}
+
+	log.Debugf("Connection between %q and %q closed.", conn.RemoteAddr(), remoteConn.RemoteAddr())
+	return nil
+}
+
+// NOTE: we assume that the `listener` has already been opened and can be used
+// to `Accept()` new connections.
+func (p *ProxyAccess) listenAndServe(address string) error {
+	// Create the listener and let it wait for new connections in a separate
+	// goroutine.
+	listener, err := net.Listen("tcp", address)
+	if err != nil {
+		return log.Errorf("net.Listen failed: %v", err)
+	}
+	defer listener.Close()
+
+	// Indefinitely accept new connections, process them concurrently.
+	for {
+		conn, err := listener.Accept() // Blocks until new connection comes
+		if err != nil {
+			return log.Errorf("listen.Accept failed: %v", err)
+		}
+
+		go p.serveConnection(conn, p.serverConfig)
+	}
+}
+
+func (p *ProxyAccess) passwordCallback(conn ssh.ConnMetadata, pass []byte) (*ssh.Permissions, error) {
+	user := conn.User()
+	log.Debugf("Login attempt from %q for user %q.", conn.RemoteAddr(), user)
+
+	fishUser, err := p.fish.UserGet(user)
+	if err != nil {
+		log.Errorf("unrecognized user %q", user)
+		return nil, fmt.Errorf("invalid access")
+	}
+
+	stringPass := string(pass)
+	ra, err := p.fish.ResourceAccessSingleUsePassword(fishUser.Name, stringPass)
+	if err != nil {
+		// NOTE: do *NOT* return the database error to the user.
+		log.Errorf("invalid access for user %q: %v", user, err)
+		return nil, fmt.Errorf("invalid access")
+	}
+
+	// Only return non-error if the username and password match.
+	if ra.Username == user && ra.Password == stringPass {
+		remoteAddr := conn.RemoteAddr()
+		// If the session is not already stored in our map, create it so that
+		// we have access to it when processing the incoming connections.
+		p.sessions.LoadOrStore(remoteAddr.String(), SessionRecord{ResourceAccessor: ra, RemoteAddr: remoteAddr})
+		return nil, nil
+	}
+
+	// Otherwise, we have failed, return error to indicate as such.
+	return nil, fmt.Errorf("invalid access")
+}
+
+func Init(fish *fish.Fish, address string) error {
+	// TODO: cannot find a way to _not_ do this.
```

Comment:
Oh wait, that's what we forgot: need to generate new `rsa private key` if it's not exist in fish_data directory.

You can pass the config parameter like in [`cmd/fish/fish.go` for cfg.TLSKey](https://github.com/adobe/aquarium-fish/blob/main/cmd/fish/fish.go#L91-L94) - create cfg.SSHKey, [set it by default to "id_rsa"](https://github.com/adobe/aquarium-fish/blob/main/lib/fish/config.go#L68-L70) and then if the file is not here - [generate new rsa key](https://stackoverflow.com/questions/64104586/use-golang-to-get-rsa-key-the-same-way-openssl-genrsa). Should be relatively easy.

Reply:
Or as in described link below: https://gist.github.com/devinodaniel/8f9b8a4f31573f428f29ec0e884e6673

Reply:
Eep, whoops!  Yes, users will _not_ have a carefully placed `id_rsa` file lying around :upside_down_face:

Implemented in https://github.com/adobe/aquarium-fish/pull/72/commits/138b303f1076de2d557cdc54a9edc06424965a1b , if we like it please resolve the conversation.  Tested it manually, it generates it the first time as expected, otherwise reloads it.  Main thing we might want to change there is the names I chose (both for the fish config `NodeSSHKey` and the actual generated files, see commit message).